File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;

class Solution {
    public int solution(int bl, int w, int[] tw) {
        List<int[]> bridge = new ArrayList<>();
        List<Integer> truckWeights = new ArrayList<>();
        
        // 트럭 무게 배열을 리스트로 변환
        for (int weight : tw) {
            truckWeights.add(weight);
        }
        
        int time = 0;
        while (true) {
            // 지나간 트럭 삭제
            if (!bridge.isEmpty() && bridge.get(0)[1] == bl) {
                bridge.remove(0);
            }
            
            // 트럭 삽입
            if (!truckWeights.isEmpty() && bridge.isEmpty()) {
                bridge.add(new int[]{truckWeights.get(0), 0});
                truckWeights.remove(0);
            }
            // 다리 길이 체크
            else if (!truckWeights.isEmpty() && bridge.size() <= bl) {
                int total = 0;
                for (int[] truck : bridge) {
                    total += truck[0];
                }
                
                // 무게 체크
                if (total + truckWeights.get(0) <= w) {
                    bridge.add(new int[]{truckWeights.get(0), 0});
                    truckWeights.remove(0);
                }
            }
            
            // 시간 경과
            time++;
            for (int[] truck : bridge) {
                truck[1]++;
            }
            
            // 종료 조건
            if (truckWeights.isEmpty() && bridge.isEmpty()) {
                break;
            }
        }
        
        return time;
    }
}

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
# bl 다리 길이, w 무게 제한, tw 트럭 무게 리스트
def solution(bl, w, tw):
    bridge = []
    time = 0
    while True:
        
        # 지나간 트럭 삭제
        if bridge and bridge[0][1] == bl:
            bridge.pop(0)
            
        # 트럭 삽입                
        if tw and not bridge:
            bridge.append([tw[0], 0])
            tw.pop(0)    
            
        # 다리 길이 체크
        elif tw and len(bridge) <= bl:
            total = 0           
            for i in bridge:
                total += i[0]
                            
            # 무게 초과
            if total + tw[0] <= w:
                bridge.append([tw[0], 0])
                tw.pop(0)
                
        # 시간 경과
        time += 1
        for i in bridge:
            i[1] += 1
                               
        if not tw and not bridge:
            break
                 
    return time